### PR TITLE
Keithh 2.4.0 beta1

### PR DESCRIPTION
**Pull request by kstevenham**
I updated the nuget package to Microsoft.ApplicationInsights.2.4.0-beta1-build07987 which touched LOTS of files. Otherwise, this check-in enables exception statistics and emits 2 example exceptions / minute for each exception for a given ProblemId.

### CODE REVIEW

File: Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs [hunk 1]
```diff
@@ -221,48 +291,210 @@ private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs fi
...
+                this.exceptionKeyValues.RwLock.EnterWriteLock();
+
+                if (this.exceptionKeyValues.ValueCache.Contains(exceptionKey) == false)
+                {
+                    this.exceptionKeyValues.ValueCache.Add(exceptionKey);
+
+                    if (exceptionTelemetry == null)
+                    {
+                        exceptionTelemetry = new ExceptionTelemetry(exception);
+                        this.telemetryClient.Initialize(exceptionTelemetry);
```
Comment (SergeyKanzhelev) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: You can do all this logic outside of WriteLock.

File: Src/WindowsServer/WindowsServer.Shared.Tests/FirstChanceExceptionStatisticsTelemetryModuleTest.cs [hunk 2]
```diff
@@ -434,6 +547,37 @@ public void FirstChanceExceptionStatisticsTelemetryModuleWasTrackedReturnsFalseF
...
+                    {
+                        throw new Exception("Second exception");
+                    }
+                    catch
+                    {
+                    }
+                }
+
+                // There should have been two unique TrackExceptions
+                Assert.Equal(2, this.items.Count);
```
Comment (kstevenham) [inline Src/WindowsServer/WindowsServer.Shared.Tests/FirstChanceExceptionStatisticsTelemetryModuleTest.cs]: I don't see any Metric items, should I?

File: Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs [hunk 3]
```diff
@@ -104,14 +127,15 @@ internal bool WasExceptionTracked(Exception exception)
...
+            ////if (!wasTracked)
+            ////{
+            ////    var innerException = exception.InnerException;
+            ////    if (innerException != null)
+            ////    {
+            ////        wasTracked = IsTracked(innerException);
+            ////    }
+            ////}
 
             if (!wasTracked && exception is AggregateException)
```
Comment (SergeyKanzhelev) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: comment this one as well then

File: Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs [hunk 4]
```diff
@@ -221,49 +299,210 @@ private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs fi
...
+            if (this.exceptionKeyValues.ValueCache.Contains(exceptionKey) == false)
+            {
+                this.exceptionKeyValues.ValueCache.Add(exceptionKey);
+
+                this.exceptionKeyValues.RwLock.ExitWriteLock();
+
+                if (exceptionTelemetry == null)
+                {
+                    exceptionTelemetry = new ExceptionTelemetry(exception);
+                    this.telemetryClient.Initialize(exceptionTelemetry);
```
Comment (SergeyKanzhelev) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: you decided to keep it inside the write lock?

Comment (kstevenham) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: It's not

File: Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs [hunk 5]
```diff
@@ -168,40 +199,79 @@ private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs fi
...
                 if (exception == null)
                 {
                     WindowsServerEventSource.Log.FirstChanceExceptionCallbackExeptionIsNull();
                     return;
                 }
 
-                WindowsServerEventSource.Log.FirstChanceExceptionCallbackCalled();
+                if (this.WasExceptionTracked(exception) == true)
+                {
+                    return;
```
Comment (kstevenham) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: This should now be resolved

File: Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs [hunk 6]
```diff
@@ -221,49 +299,210 @@ private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs fi
...
+
+                this.exceptionKeyValues.RwLock.ExitWriteLock();
+
+                if (exceptionTelemetry == null)
+                {
+                    exceptionTelemetry = new ExceptionTelemetry(exception);
+                    this.telemetryClient.Initialize(exceptionTelemetry);
+                }
+
+                StackTrace st = new StackTrace(3, true);
```
Comment (Dmitry-Matveev) [inline Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs]: Is there a way to ensure that no in-lining or other optimization happens and skipping 3 frames always leads to the consistent result? The previous methods do not seem to satisfy any of the in-lining/optimization requirements, so this should work as is unless those requirements change in CLR.

**Comment by lmolkova** [comment]
It seems that after this change TestAsmx_CheckExceptionAndRequestCollectedForResourceNotFound Functional test does not work, because it expect just one exception to be tracked and now there are multiple.
@kstevenham Could you please check it?

**Pull request merged**